### PR TITLE
jamie/scale cmd

### PR DESCRIPTION
**Pull request by jamiealquiza**
# Changes
Adds the `scale` sub-command. Docs, testing pending.

`scale` is an equivalent to running `rebalance` with the following differences:
- Broker additions are required
- The `--storage-threshold` parameter is removed; all non-new brokers are automatically targeted for selection
- Increases to the storage range spread aren't counted as a warning

Other changes:
- `scale` is very similar to `rebalance`; `rebalance_steps.go` has been dropped and many functions have been made generic for sharing between `scale` and `rebalance`
- Some minor cleanup

Further cleanup and refactoring will be in a dedicated follow up PR.

# Examples

The new `scale` command creating the same results as a `rebalance`, but without the parameter wizardry:

```
$ topicmappr rebalance --topics test --brokers -2 --storage-threshold 0 --ignore-warns --out-file a
$ ./topicmappr scale --topics test  --brokers -2 --out-file b
$ $ diff a.json b.json ; echo $?
0
```

Particularly, `scale` doesn't consider storage range increases a problem since its objective is to increase broker counts, whereas `rebalance` is focused on improving storage distribution:

```
$ ./topicmappr scale --topics test --brokers -2 | grep -A1 WARN 
WARN:
  [none]
$ topicmappr rebalance --topics test --brokers -2 --storage-threshold 0 | grep -A1 WARN
WARN:
  broker free storage range increased
```

`scale` requires additional brokers, unlike `rebalance`:
```
$ ./topicmappr scale --topics test --brokers -1
Topics:
  test
Validating broker list:
  [ERROR] scale requires additional brokers
```

### CODE REVIEW

File: cmd/topicmappr/commands/output.go [hunk 1]
```diff
@@ -195,7 +201,7 @@ func printBrokerAssignmentStats(cmd *cobra.Command, pm1, pm2 *kafkazk.PartitionM
 			newStorage := bm2[id].StorageFree / div
 
 			// Skip reporting non-changes when using rebalance.
```
Comment (scanterog) [inline cmd/topicmappr/commands/output.go:203]: shall we just remove these comments?

File: cmd/topicmappr/commands/scale.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
+// The scale command is mostly the same as rebalance, so several common functions
+// are shared from the rebalance_steps file.
```
Comment (scanterog) [inline cmd/topicmappr/commands/scale.go:2]: nit: update this now that rebalance_steps is gone.

File: cmd/topicmappr/commands/scale.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	})
+
+	// Chose the results with the lowest range.
+	m := resultsByRange[0]
+	partitionMapOut, brokersOut, relos := m.partitionMap, m.brokers, m.relocations
+
+	// Print parameters used for scale decisions.
+	printReassignmentParams(cmd, resultsByRange, brokersIn, m.tolerance)
+
+	// Optimize leaders.
```
Comment (scanterog) [inline cmd/topicmappr/commands/scale.go:136]: is this expected to have in a scale operation?

**Review by scanterog — APPROVED** [review]

File: cmd/topicmappr/commands/scale.go (hunk 3 above)
Comment (jamiealquiza) [reply cmd/topicmappr/commands/scale.go:134]: It's optional, so just keeping it consistent with the other commands (in that you can decide to also perform a leadership optimization whether a scale, rebalance, or rebuild)

**Pull request merged**